### PR TITLE
Implementation of the `Profile` listings page integration  #37

### PR DESCRIPTION
**Pull request by fetahokey**
### What this PR does (required):
- Modified `Profile` **model** and **controllers** to match listings requirements 
- Integrated Listings Page UI
- Display the availability on the listing items 

### Information needed to test this feature :
-  Mock-data on the database for the documents `User`, `Profile`
-  Users roles should be set to 'sitter'
-  Start the app and navigate to  `/listing`

### Issues with the current functionality :
- `rating` field is hardcoded (random value)

### Features:
- [x] #37
- [ ] Pagintation
- [ ] Lazy loading

### CODE REVIEW

File: client/src/pages/Listing/Listing.tsx [hunk 1]
```diff
@@ -1,77 +1,91 @@
...
+  const [profiles, setProfiles] = useState<Profile[]>([]);
+  const [dateIn, setDateIn] = useState<number>(TODAY);
+  const [dateOff, setDateOff] = useState<number>(TODAY);
+  const [searchString, setSearchString] = useState<string>('');
+  const dateInRef = useRef<any>();
+  const dateOffRef = useRef<any>();
+  const searchRef = useRef<any>();
+
+  const handleFilterChange = () => {
+    setFilter(!filter);
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:30]:
```suggestion
    setFilter(prevFilter => !prevFilter);
```

It is a good practice to pass a function that takes the current state as the first argument!

File: client/src/pages/Listing/Listing.tsx [hunk 2]
```diff
@@ -1,77 +1,91 @@
...
   const { root, title, search, dateInOff, iconStyle } = useStyles();
+  const TODAY = new Date().getDay();
+  const [loading, setLoading] = useState<undefined | boolean>(undefined);
+  const [filter, setFilter] = useState<boolean>(false);
+  const { updateSnackBarMessage } = useSnackBar();
+  const [profiles, setProfiles] = useState<Profile[]>([]);
+  const [dateIn, setDateIn] = useState<number>(TODAY);
+  const [dateOff, setDateOff] = useState<number>(TODAY);
+  const [searchString, setSearchString] = useState<string>('');
+  const dateInRef = useRef<any>();
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:25]: Please avoid using 'any' as a type :)

File: client/src/pages/Listing/Listing.tsx [hunk 3]
```diff
@@ -105,44 +121,70 @@ const Listing = (): JSX.Element => {
...
-                image={img}
-                fullName={fullname}
-                rating={rating as rate}
-                slogan={slogan}
-                description={description}
-                location={location}
-                price={price}
-              />
-            ))}
+            {!loading && profiles.length > 0 ? (
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:156]:
```suggestion
            {!loading && !profiles.length ? (
```

File: client/src/pages/Listing/Listing.tsx [hunk 4]
```diff
@@ -105,44 +121,70 @@ const Listing = (): JSX.Element => {
...
+              profiles.map(
+                ({ _id, address, description, firstName, gender, lastName, phoneNumber, rate, availability }) => (
+                  <ListingItem
+                    key={_id}
+                    image={'img'}
+                    fullName={getFullName(firstName, lastName)}
+                    rating={Math.floor(Math.random() * 5) as rate}
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:163]: 👍🏽

File: client/src/pages/Listing/Listing.tsx [hunk 5]
```diff
@@ -105,44 +121,70 @@ const Listing = (): JSX.Element => {
...
+                    slogan={'slogan'}
+                    description={description}
+                    location={address}
+                    price={rate || 0}
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:167]: **Suggestion:** maybe consider adding a default value to the 'price' field in the relevant model

File: client/src/pages/Listing/ListingItem.tsx [hunk 6]
```diff
@@ -49,11 +79,56 @@ export const ListingItem = ({ image, fullName, rating, slogan, description, loca
...
               </Typography>
-              <Typography variant="subtitle1">{slogan}</Typography>
+
               <Rating name="read-only" value={rating as number} readOnly />
               <Typography className={descriptionStyle} align="center">
                 {description}
               </Typography>
+              <Box pt={2}>
+                <Divider light />
+                <Chip
```
Comment (rajivtitus) [inline client/src/pages/Listing/ListingItem.tsx:89]: Good thinking adding the available days as a Chip to each individual card. However, it does seem a bit cluttered, I would recommend adding these chips to the Profile Details page instead. Thoughts? :)

File: server/controllers/profile.js [hunk 7]
```diff
@@ -1,63 +1,108 @@
...
-
+  let { filter = false, dateIn = 0, dateOff = 0, search = "" } = req.query;
+  filter = stringToBoolean(filter);
+  dateIn = parseInt(dateIn);
+  dateOff = parseInt(dateOff);
 
+  const profiles = await Profile.find()
+    .populate({
+      path: "user",
+      select: { password: 0 },
```
Comment (rajivtitus) [inline server/controllers/profile.js:73]: 👍🏽

File: client/src/pages/Listing/Listing.tsx [hunk 8]
```diff
@@ -1,77 +1,91 @@
...
 const Listing = (): JSX.Element => {
   const { root, title, search, dateInOff, iconStyle } = useStyles();
+  const TODAY = new Date().getDay();
+  const [loading, setLoading] = useState<undefined | boolean>(undefined);
+  const [filter, setFilter] = useState<boolean>(false);
+  const { updateSnackBarMessage } = useSnackBar();
+  const [profiles, setProfiles] = useState<Profile[]>([]);
+  const [dateIn, setDateIn] = useState<number>(TODAY);
+  const [dateOff, setDateOff] = useState<number>(TODAY);
+  const [searchString, setSearchString] = useState<string>('');
```
Comment (rajivtitus) [inline client/src/pages/Listing/Listing.tsx:24]: **Bonus:** I would recommend looking into debouncing so that we don't overload the server 😄

**Review by rajivtitus — APPROVED** [review]
Good work with this! Just left you a few comments, please have a look and address before merging. Thanks!

File: client/src/pages/Listing/Listing.tsx (hunk 2 above)
Comment (fetahokey) [reply client/src/pages/Listing/Listing.tsx:25]: done ✔️ 

`HTMLInputElement`
`React.ChangeEvent<HTMLInputElement>`
types 
instead if `any`

File: client/src/pages/Listing/Listing.tsx (hunk 1 above)
Comment (fetahokey) [reply client/src/pages/Listing/Listing.tsx:30]: done ✔️

File: client/src/pages/Listing/Listing.tsx (hunk 3 above)
Comment (fetahokey) [reply client/src/pages/Listing/Listing.tsx:156]: Applied  🖌️

File: client/src/pages/Listing/ListingItem.tsx (hunk 6 above)
Comment (fetahokey) [reply client/src/pages/Listing/ListingItem.tsx:89]: 🙏 
>Please, updated arranged && organized in an orderly way. 😄 
>image

Comment (rajivtitus) [reply client/src/pages/Listing/ListingItem.tsx:89]: This looks a lot better, thanks for changing it 😆

File: client/src/pages/Listing/Listing.tsx (hunk 8 above)
Comment (fetahokey) [reply client/src/pages/Listing/Listing.tsx:24]: done ✔️ 
 ```
   const [debouncedSearch] = useDebounce({ filter, dateIn, dateOff, searchString }, 800);
   ____ 
 listProfiles({ ...debouncedSearch }).then((response) => {
        if (response.error) { ...

**Comment by fetahokey** [comment]
> Good work with this! Just left you a few comments, please have a look and address before merging. Thanks!
😄  Thank you

**Pull request merged**